src/System.Collections/tests/Generic/Dictionary/Dictionary.Tests.cs

```diff
@@ -405,6 +407,53 @@ private static IEnumerable<object[]> GetCopyConstructorData<T>(Func<int, T> keyV
             return dict;
         }
 
+        [Fact]
+        public void ComparerSerialization()
+        {
+            // Strings switch between randomized and non-randomized comparers, 
+            // however this should never be observable externally.
+            TestComparerSerialization(EqualityComparer<string>.Default);
+            // OrdinalCaseSensitiveComparer is internal and (de)serializes as OrdinalComparer
+            TestComparerSerialization(StringComparer.Ordinal, "System.OrdinalComparer");
+            // OrdinalIgnoreCaseComparer is internal and (de)serializes as OrdinalComparer
+            TestComparerSerialization(StringComparer.OrdinalIgnoreCase, "System.OrdinalComparer");
+            TestComparerSerialization(StringComparer.CurrentCulture);
+            TestComparerSerialization(StringComparer.CurrentCultureIgnoreCase);
+            TestComparerSerialization(StringComparer.InvariantCulture);
+            TestComparerSerialization(StringComparer.InvariantCultureIgnoreCase);
+
+            // Check other types while here, IEquatable valuetype, nullable valuetype, and non IEquatable object
+            TestComparerSerialization(EqualityComparer<int>.Default);
+            TestComparerSerialization(EqualityComparer<int?>.Default);
+            TestComparerSerialization(EqualityComparer<object>.Default);
+        }
+
+        private static void TestComparerSerialization<T>(IEqualityComparer<T> equalityComparer, string internalTypeName = null)
+        {
+            var bf = new BinaryFormatter();
+            var s = new MemoryStream();
+
+            var dict = new Dictionary<T, T>(equalityComparer);
+
+            Assert.IsType(equalityComparer.GetType(), dict.Comparer);
```

Comment:
Does dictionary not guarantee `Assert.Same(equalityComparer, dict.Comparer);`?

Reply:
Guarantee might be a strong word :)

They would work pre-serilization because they are the same statics instances in use (even though it swaps them out, it swaps them back to the same).

`StringComparer.Ordinal` and `StringComparer.OrdinalIgnoreCase` changing type after serialization surprised me and none of the instances will be same as the static ones post-serialization.

Reply:
> They would work pre-serilization because they are the same statics instances in use (even though it swaps them out, it swaps them back to the same).

Right, that's what I meant; this Assert.IsType is validating the types are the same prior to serialization, but it seems like we should tighten that to use Assert.Same here.  We'd keep the same asserts you have for the post-validation.

Reply:
Changed